magda-gateway/src/createBaseProxy.ts

```diff
@@ -75,5 +84,42 @@ export default function createBaseProxy(): httpProxy {
         });
     });
 
+    proxy.on("proxyReq", async function(proxyReq, req, res) {
+        proxyReq.setHeader(MAGDA_TENANT_ID_HEADER, "undefined");
+
+        if (multiTenantsMode === true) {
+            const host = req.headers.host;
```

Comment:
You should actually use [req.hostname](https://expressjs.com/en/api.html#req.hostname), which will automatically use `X-Forwarded-Host` depending on the configuration of the `trust proxy` setting (see https://expressjs.com/en/guide/behind-proxies.html).

Reply:
Do I understand this correctly:
If the request is from a trusted proxy, we get the host from the first X-Forwarded-Host header whose value is allowed to be set by client or proxy. Otherwise no proxy is involved and proceeds as normal. 
I am wondering what the use cases are. Will the gateway be put behind a proxy?

Reply:
Reverse proxies are really common, e.g. load balancers, CloudFront, maybe even ingress is done as a reverse proxy on GKE (or maybe not, I'm not sure). But you don't need to think about any of that, just use `req.hostname` instead of getting the header directly and express takes care of the rest.

Reply:
Ok. Will fix it.

Reply:
Actually I think the request here might be the native nodejs version rather than the express version, so it won't have the fancy `hostname` stuff :(

Reply:
Kevin helped me figure it out last week. I decided to use type cast quick solution. Please see the updated version.

Reply:
done